### PR TITLE
PLT-2989 Incoming webhooks available to all channels user is member of

### PR DESCRIPTION
**Pull request by DavidLu1997**
Removed the check for OPEN_CHANNEL, now incoming/outgoing webhooks can be directed at any channel that is not a DM channel.

Can also change the check to allow DM channels.

### CODE REVIEW

**Comment by enahum** [comment]
+1

**Comment by coreyhulen** [comment]
This should only be for incoming right? Outgoing should not see private channels?

**Comment by DavidLu1997** [comment]
Changed so that outgoing does not see private

**Comment by hmhealey** [comment]
This doesn't appear to fix the issue that the user can't add incoming webhooks to public channels that they aren't a member of

**Comment by DavidLu1997** [comment]
```
Build was abortedBuild was aborted
Aborted by Christopher Speller
[TemplateProject] Starting publishers from: mattermost-platform-template
```
@crspeller rebuild please? :)

**Comment by crspeller** [comment]
@DavidLu1997 You should be able to do it yourself. I aborted it because it was in-front of the build to fix pre-release.

**Comment by DavidLu1997** [comment]
Did not see the rebuild option, thank you :)

File: webapp/components/channel_select.jsx [hunk 1]
```diff
@@ -6,43 +6,65 @@ import React from 'react';
...
     }
 
     handleChannelChange() {
         this.setState({
-            channels: ChannelStore.getAll()
+            channels: ChannelStore.getAll().concat(ChannelStore.getMoreAll()).sort(this.compareByDisplayName)
         });
     }
 
+    compareByDisplayName(a, b) {
```
Comment (crspeller) [inline webapp/components/channel_select.jsx]: Compare what by display name? The argument names of `a` and `b` certainly don't help.

File: webapp/components/channel_select.jsx [hunk 2]
```diff
@@ -6,43 +6,65 @@ import React from 'react';
...
     componentWillUnmount() {
         ChannelStore.removeChangeListener(this.handleChannelChange);
+        ChannelStore.removeMoreChangeListener(this.onListenerChange);
+    }
+
+    onListenerChange() {
+        var newState = {
+            channels: ChannelStore.getAll().concat(ChannelStore.getMoreAll()).sort(this.compareByDisplayName)
+        };
+        if (!Utils.areObjectsEqual(newState.channels, this.state.channels)) {
```
Comment (crspeller) [inline webapp/components/channel_select.jsx]: The proper way to do this is to write a shouldComponentUpdate function. You can see examples of this throughout the code. (probably also examples of us doing it wrong like this :P)

File: webapp/components/channel_select.jsx [hunk 3]
```diff
@@ -77,3 +117,9 @@ export default class ChannelSelect extends React.Component {
         );
     }
 }
+
+ChannelSelect.defaultProps = {
```
Comment (crspeller) [inline webapp/components/channel_select.jsx]: We use a get function for the propTypes but this for default props. We should be consistent.

**Comment by DavidLu1997** [comment]
This should be fixed now

File: webapp/components/channel_select.jsx [hunk 4]
```diff
@@ -37,12 +51,20 @@ export default class ChannelSelect extends React.Component {
         ChannelStore.removeChangeListener(this.handleChannelChange);
     }
 
+    shouldComponentUpdate(nextProps, nextState) {
+        return !Utils.areObjectsEqual(nextState.channels, this.state.channels);
+    }
```
Comment (jwilander) [inline webapp/components/channel_select.jsx]: Please remove `shouldComponentUpdate()`

File: webapp/components/channel_select.jsx [hunk 5]
```diff
@@ -6,28 +6,42 @@ import React from 'react';
...
 
         this.handleChannelChange = this.handleChannelChange.bind(this);
+        this.compareByDisplayName = this.compareByDisplayName.bind(this);
 
         this.state = {
             channels: []
         };
     }
 
     componentWillMount() {
```
Comment (jwilander) [inline webapp/components/channel_select.jsx]: Please move all contents to constructor and remove `componentWillMount()`

File: webapp/components/channel_select.jsx [hunk 6]
```diff
@@ -6,28 +6,42 @@ import React from 'react';
...
+        AsyncClient.getMoreChannels(true);
         this.setState({
-            channels: ChannelStore.getAll()
+            channels: ChannelStore.getAll().sort(this.compareByDisplayName)
```
Comment (jwilander) [inline webapp/components/channel_select.jsx]: Don't need this setState, and please add intialization in constructor instead

**Comment by jwilander** [comment]
+1

**Comment by crspeller** [comment]
+1

**Pull request merged**